### PR TITLE
Refactor the fetch shader creation

### PR DESCRIPTION
**Pull request by s-perron**
Splits up the function that generates the fetch shader into smaller functions.

### CODE REVIEW

File: lgc/elfLinker/FetchShader.cpp [hunk 1]
```diff
@@ -114,42 +136,104 @@ Module *FetchShader::generate() {
...
+  default:
+    llvm_unreachable("Unexpected special user data or shader input");
+  }
+  return nullptr;
+}
+
+// =====================================================================================================================
+// @param call : A call to the InstanceId shader input builtin.
+// @returns : The value of the InstanceId.  All new code will be place at the start of the function containing call.
+Value *FetchShader::getReplacementForInstanceIdBuildIn(CallInst *call) const {
```
Comment (dstutt) [inline lgc/elfLinker/FetchShader.cpp:172]: Nit BuildIn vs BuiltIn (maybe this is deliberate though?)

**Review by dstutt — APPROVED** [review]
LGTM with minor nit (possibly)
Might be good to get TimR to comment as well

**Comment by s-perron** [comment]
V2: Fixed the typo @dstutt mentioned.  I also removed changes that were suppose to be part of #1743.

File: lgc/elfLinker/FetchShader.cpp (hunk 1 above)
Comment (s-perron) [reply lgc/elfLinker/FetchShader.cpp]: Fixed.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 8dfd2e52673886b3219f3137a93eb26162cb205a
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 39f8940199e60c44d4211cf8165dfd12876316fa</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 2483d46fa27c30502e497ea169ee53b142e9fa06</li>
<li>SPVGEN: dev fa39b110650bc72fd7fda54af621e69306036888</li>
<li>XGL: dev 6d693c7342bb7f18318e80c1da53061dfcd48262</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 2a87cb67e0180322832e722db8a4c53a5fc4e867</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/203169)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 38237/67723 (56.5%)</li>
<li>Failed: 0/67723 (0.0%)</li>
<li>Not Supported: 29486/67723 (43.5%)</li><li>Warnings: 0/67723 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 38075/67723 (56.2%)</li>
<li>Failed: 0/67723 (0.0%)</li>
<li>Not Supported: 29648/67723 (43.8%)</li><li>Warnings: 0/67723 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 39235/67723 (57.9%)</li>
<li>Failed: 0/67723 (0.0%)</li>
<li>Not Supported: 28488/67723 (42.1%)</li><li>Warnings: 0/67723 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit bcab1820e4df3e30a8f6d522d71bf46c8b748c9e
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 39f8940199e60c44d4211cf8165dfd12876316fa</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 2483d46fa27c30502e497ea169ee53b142e9fa06</li>
<li>SPVGEN: dev fa39b110650bc72fd7fda54af621e69306036888</li>
<li>XGL: dev 6d693c7342bb7f18318e80c1da53061dfcd48262</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 2a87cb67e0180322832e722db8a4c53a5fc4e867</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/202608)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 38237/67536 (56.6%)</li>
<li>Failed: 0/67536 (0.0%)</li>
<li>Not Supported: 29299/67536 (43.4%)</li><li>Warnings: 0/67536 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 38075/67536 (56.4%)</li>
<li>Failed: 0/67536 (0.0%)</li>
<li>Not Supported: 29461/67536 (43.6%)</li><li>Warnings: 0/67536 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 39235/67536 (58.1%)</li>
<li>Failed: 0/67536 (0.0%)</li>
<li>Not Supported: 28301/67536 (41.9%)</li><li>Warnings: 0/67536 (0.0%)</li>
</ul></details>


</ul></details>

File: lgc/elfLinker/FetchShader.cpp [hunk 2]
```diff
@@ -114,42 +132,74 @@ Module *FetchShader::generate() {
...
+Value *FetchShader::getReplacementForInstanceIdBuiltIn(CallInst *call) const {
+  Function *callerFunction = call->getParent()->getParent();
+  return getVgprArgument(m_vsEntryRegInfo.instanceId, callerFunction);
+}
+
+// =====================================================================================================================
+// @param call : A call to the VertexId shader input builtin.
+// @returns : The value of the VertexId.  All new code will be place at the start of the function containing call.
+Value *FetchShader::getReplacementForVertexIdBuiltIn(CallInst *call) const {
+  Function *callerFunction = call->getParent()->getParent();
```
Comment (kuhar) [inline lgc/elfLinker/FetchShader.cpp:177]: also here

File: lgc/elfLinker/FetchShader.cpp [hunk 3]
```diff
@@ -104,8 +114,16 @@ Module *FetchShader::generate() {
     }
   }
   ret->setOperand(0, result);
+}
 
-  // Hook up the inputs (vertex buffer, base vertex, base instance, vertex ID, instance ID). The fetchVertex calls
+// =====================================================================================================================
+// Replaces calls to the shader input builtins in fetchFunc with code that will get the appropriate values from the
+// arguments.
+void FetchShader::replaceShaderInputBuiltInFunctions(Function *fetchFunc) const {
```
Comment (kuhar) [inline lgc/elfLinker/FetchShader.cpp:122]: nit: `@param` is missing

**Review by kuhar — COMMENTED** [review]
Just a few nits, LGTM otherwise.

File: lgc/elfLinker/FetchShader.cpp [hunk 4]
```diff
@@ -114,42 +132,74 @@ Module *FetchShader::generate() {
...
+    llvm_unreachable("Unexpected special user data or shader input");
+  }
+  return nullptr;
+}
+
+// =====================================================================================================================
+// @param call : A call to the InstanceId shader input builtin.
+// @returns : The value of the InstanceId.  All new code will be place at the start of the function containing call.
+Value *FetchShader::getReplacementForInstanceIdBuiltIn(CallInst *call) const {
+  Function *callerFunction = call->getParent()->getParent();
```
[the remark this answers is not in the document]
Comment (s-perron) [reply lgc/elfLinker/FetchShader.cpp]: I was looking for it, but for some reason I did not see it.  Thanks.

File: lgc/elfLinker/FetchShader.h [hunk 5]
```diff
@@ -73,6 +74,14 @@ class FetchShader : public GlueShader {
...
 private:
   llvm::Function *createFetchFunc();
+  void generateFetchShaderBody(llvm::Function *fetchFunc);
+
+  void replaceShaderInputBuiltInFunctions(llvm::Function *fetchFunc) const;
+  llvm::Value *getReplacementForVertexBufferTableBuiltIn(llvm::CallInst *call) const;
+  llvm::Value *getReplacementForInputBuiltIn(llvm::CallInst *call) const;
+  llvm::Value *getReplacementForVertexIdBuiltIn(llvm::CallInst *call) const;
+  llvm::Value *getReplacementForInstanceIdBuiltIn(llvm::CallInst *call) const;
+  llvm::Value *getVgprArgument(unsigned int vgpr, llvm::Function *function) const;
```
Comment (kuhar) [inline lgc/elfLinker/FetchShader.h:84]: nit: unsigned int -> unsigned

File: lgc/elfLinker/FetchShader.cpp [hunk 6]
```diff
@@ -114,42 +134,74 @@ Module *FetchShader::generate() {
...
+Value *FetchShader::getReplacementForVertexIdBuiltIn(CallInst *call) const {
+  Function *callerFunction = call->getFunction();
+  return getVgprArgument(m_vsEntryRegInfo.vertexId, callerFunction);
+}
+
+// =====================================================================================================================
+// @param vgpr : The vgpr number from which the value should come.
+// @param function : The function from which to get the argument.
+// @returns : The value of the argument in the function that corresponds to the given vgpr.
+Value *FetchShader::getVgprArgument(unsigned int vgpr, Function *function) const {
```
Comment (kuhar) [inline lgc/elfLinker/FetchShader.cpp:187]: nit: unsigned int -> unsigned

**Review by kuhar — APPROVED** [review]

**Review by kuhar — DISMISSED** [review]

**Comment by amdvlk-admin** [comment]
### Test summary for commit 0b658b56bd7dda392fdda78d113ab5d2f58a3f25
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 39f8940199e60c44d4211cf8165dfd12876316fa</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 2483d46fa27c30502e497ea169ee53b142e9fa06</li>
<li>SPVGEN: dev fa39b110650bc72fd7fda54af621e69306036888</li>
<li>XGL: dev 6d693c7342bb7f18318e80c1da53061dfcd48262</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 2a87cb67e0180322832e722db8a4c53a5fc4e867</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/202608)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 38237/67536 (56.6%)</li>
<li>Failed: 0/67536 (0.0%)</li>
<li>Not Supported: 29299/67536 (43.4%)</li><li>Warnings: 0/67536 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 38075/67536 (56.4%)</li>
<li>Failed: 0/67536 (0.0%)</li>
<li>Not Supported: 29461/67536 (43.6%)</li><li>Warnings: 0/67536 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 39235/67536 (58.1%)</li>
<li>Failed: 0/67536 (0.0%)</li>
<li>Not Supported: 28301/67536 (41.9%)</li><li>Warnings: 0/67536 (0.0%)</li>
</ul></details>


</ul></details>

File: lgc/elfLinker/FetchShader.cpp [hunk 7]
```diff
@@ -114,42 +134,74 @@ Module *FetchShader::generate() {
...
+// =====================================================================================================================
+// @param call : A call to the VertexId shader input builtin.
+// @returns : The value of the VertexId.  All new code will be place at the start of the function containing call.
+Value *FetchShader::getReplacementForVertexIdBuiltIn(CallInst *call) const {
+  Function *callerFunction = call->getFunction();
+  return getVgprArgument(m_vsEntryRegInfo.vertexId, callerFunction);
+}
+
+// =====================================================================================================================
+// @param vgpr : The vgpr number from which the value should come.
```
Comment (amdrexu) [inline lgc/elfLinker/FetchShader.cpp:184]: Prefer to use `VGPR` in comments.

File: lgc/elfLinker/FetchShader.cpp [hunk 8]
```diff
@@ -114,42 +134,74 @@ Module *FetchShader::generate() {
...
+// @param function : The function from which to get the argument.
+// @returns : The value of the argument in the function that corresponds to the given vgpr.
```
Comment (amdrexu) [inline lgc/elfLinker/FetchShader.cpp:186]: The same as above.

File: lgc/elfLinker/FetchShader.cpp [hunk 9]
```diff
@@ -83,12 +87,18 @@ StringRef FetchShader::getMainShaderName() {
...
   // Create the function.
   Function *fetchFunc = createFetchFunc();
+  generateFetchShaderBody(fetchFunc);
+  replaceShaderInputBuiltInFunctions(fetchFunc);
+  return fetchFunc->getParent();
+}
 
-  // Process each vertex input.
+// =====================================================================================================================
+// Generate the body of the fetch function using the shader input builtins to access the inputs to the shader.
```
Comment (amdrexu) [inline lgc/elfLinker/FetchShader.cpp:96]: Missing parameter comments @fetchFunc

File: lgc/elfLinker/FetchShader.cpp [hunk 10]
```diff
@@ -114,42 +134,74 @@ Module *FetchShader::generate() {
...
+        replacement = getReplacementForInputBuiltIn(call);
         call->replaceAllUsesWith(replacement);
         call->eraseFromParent();
       }
     }
   }
+}
 
-  return fetchFunc->getParent();
+// =====================================================================================================================
```
Comment (amdrexu) [inline lgc/elfLinker/FetchShader.cpp:145]: Add function description comments.

File: lgc/elfLinker/FetchShader.cpp [hunk 11]
```diff
@@ -114,42 +134,74 @@ Module *FetchShader::generate() {
...
+    return getReplacementForVertexIdBuiltIn(call);
+  case static_cast<unsigned>(ShaderInput::InstanceId):
+    return getReplacementForInstanceIdBuiltIn(call);
+  default:
+    llvm_unreachable("Unexpected special user data or shader input");
+  }
+  return nullptr;
+}
+
+// =====================================================================================================================
```
Comment (amdrexu) [inline lgc/elfLinker/FetchShader.cpp:167]: The same as above.

File: lgc/elfLinker/FetchShader.cpp [hunk 12]
```diff
@@ -114,42 +134,74 @@ Module *FetchShader::generate() {
...
+// @param call : A call to the InstanceId shader input builtin.
+// @returns : The value of the InstanceId.  All new code will be place at the start of the function containing call.
+Value *FetchShader::getReplacementForInstanceIdBuiltIn(CallInst *call) const {
+  Function *callerFunction = call->getFunction();
+  return getVgprArgument(m_vsEntryRegInfo.instanceId, callerFunction);
+}
+
+// =====================================================================================================================
```
Comment (amdrexu) [inline lgc/elfLinker/FetchShader.cpp:175]: The same as above.

File: lgc/elfLinker/FetchShader.cpp [hunk 13]
```diff
@@ -114,42 +134,74 @@ Module *FetchShader::generate() {
...
+// @param vgpr : The vgpr number from which the value should come.
+// @param function : The function from which to get the argument.
+// @returns : The value of the argument in the function that corresponds to the given vgpr.
+Value *FetchShader::getVgprArgument(unsigned vgpr, Function *function) const {
+  BuilderBase builder(&*function->front().getFirstInsertionPt());
+  Argument *vertexId = function->getArg(m_vsEntryRegInfo.sgprCount + vgpr);
+  return builder.CreateBitCast(vertexId, builder.getInt32Ty());
+}
+
+// =====================================================================================================================
```
Comment (amdrexu) [inline lgc/elfLinker/FetchShader.cpp:193]: The same as above.

**Review by amdrexu — COMMENTED** [review]
LGTM. Just some nits.

File: lgc/elfLinker/FetchShader.cpp (hunk 10 above)
Comment (s-perron) [reply lgc/elfLinker/FetchShader.cpp:145]: I'll do it, but I have a problem with how much extra work our coding standards add for no extra value.  What extra information will the function description add that is not covered by the "@return".  If I add the description, but not the "@return", then I get a comment that the code standards document says I have to have the "@return".  On top of that, these comments are often just repeating the same information that is in the function name.

Comment (amdrexu) [reply lgc/elfLinker/FetchShader.cpp:147]: I think AMD follows this style in other projects:
```
// =====
// Function descriptions
//
// NOTE: Extra comments that need special attention.
//
// @param X :
// @param Y :
// @return ...
```

**Comment by amdvlk-admin** [comment]
### Test summary for commit e5a982b78a87f143890a5eb6f8f5d43cbf0a8979
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 39f8940199e60c44d4211cf8165dfd12876316fa</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 2483d46fa27c30502e497ea169ee53b142e9fa06</li>
<li>SPVGEN: dev fa39b110650bc72fd7fda54af621e69306036888</li>
<li>XGL: dev 6d693c7342bb7f18318e80c1da53061dfcd48262</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 2a87cb67e0180322832e722db8a4c53a5fc4e867</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/202608)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 38237/67536 (56.6%)</li>
<li>Failed: 0/67536 (0.0%)</li>
<li>Not Supported: 29299/67536 (43.4%)</li><li>Warnings: 0/67536 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 38075/67536 (56.4%)</li>
<li>Failed: 0/67536 (0.0%)</li>
<li>Not Supported: 29461/67536 (43.6%)</li><li>Warnings: 0/67536 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 39235/67536 (58.1%)</li>
<li>Failed: 0/67536 (0.0%)</li>
<li>Not Supported: 28301/67536 (41.9%)</li><li>Warnings: 0/67536 (0.0%)</li>
</ul></details>


</ul></details>

**Pull request merged**